### PR TITLE
Create the ClientPool decorator that uses an ε-greedy algorithm to distribute load

### PR DESCRIPTION
**Pull request by fgrosse**
Still missing a proper unit test but I have a working proof of concept with multiple servers that change their response duration over time.

### CODE REVIEW

File: client_pool.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+// HTTP requests among multiple hosts. The pool transparently distributes the requests
+// among the hosts taking the individual request durations and failures into account.
+// A more detailed discussion of the underlying algorithm can be found at
+// https://godoc.org/github.com/bitly/go-hostpool#NewEpsilonGreedy
+//
+// See also https://en.wikipedia.org/wiki/Epsilon-greedy_strategy
+func ClientPool(hosts []string, decayDuration time.Duration, valueCalculator hostpool.EpsilonValueCalculator) Decorator {
+	pool := hostpool.NewEpsilonGreedy(hosts, decayDuration, valueCalculator)
+	clients := make(map[string]Client)
+	mux := &sync.Mutex{}
```
Comment (pschultz) [inline client_pool.go]: Can we call this mu or lock instead? I find this confusing because mux usually means multiplexer, as in http.ServeMux.

**Comment by pschultz** [comment]
Since ClientPool is implemented in terms of Scoped, it doesn't take hostnames as an argument but base urls. That means that requests with already absolute urls are not really distributed (Scoped does nothing in this case). 

So basically, ClientPool and Scoped are mutually exclusive. I'm having trouble deciding if I like that or not. An alternative would be to replace just the Host part of the request, but that is awkward for client code, because it still has to supply the scheme. I dunno. Just wanted to mention it.

**Comment by fgrosse** [comment]
Not only would the client code have to supply the scheme but also the port. I believe  this decorator is more useful if we do not make any assumptions like that.

If its clear that we require complete URLs this decorator really is equal to `Scoped` but for multiple domains. I dont see any usecase where combining both makes any sense.

Maybe this can just be resolved with renaming the variable `hosts`?

**Comment by pschultz** [comment]
I came to the same conclusion after thinking about it. If a client is decorated with both ClientPool and Scoped, the outer decorator will never have any effect.

In the future, we should have another ClientPool implementation that allows updating the host list to facilitate service discovery. I have an idea how to that. I don't want to increase the scope of the PR now, but I will rename ClientPool to StaticClientPool so we don't have a BC break down the line; we can then add DynamicClientPool later.

I'll also rename the urls parameter to baseURLs to make it consistent with Scoped.

**Pull request merged**